[mlir][OpenMP] - MLIR to LLVMIR translation support for delayed privatization in `omp.target` ops.

bhandarkar-pranav opened the pull request:
This patch adds support to translate the `private` clause on `omp.target` ops from MLIR to LLVMIR. This first cut only handles non-allocatables. Also, this is for delayed privatization.

bhandarkar-pranav committed:
[mlir][OpenMP] - Implement lowering from MLIR to LLVMIR for private clause
on target constructs

This patch adds support to translate the `private` clause on `omp.target`
ops from MLIR to LLVMIR. This first cut only handles non-allocatable.
Also, this is for delayed privatization.

Dinistro commented on mlir/test/Target/LLVMIR/openmp-target-private.mlir:
Nit: Please ensure that the IR is properly formatted and is stripped to the absolute minimum to test this feature.

Dinistro commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -1356,6 +1356,38 @@ class OmpParallelOpConversionManager {
 };
 
+namespace {
+omp::PrivateClauseOp findPrivatizer(Operation *from, SymbolRefAttr symbolName) {
```
Nit: MLIR prefers static functions over functions in anonymous namespaces. Also, please comment functions.

Dinistro commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -3435,6 +3447,58 @@ convertOmpTarget(Operation &opInst, llvm::IRBuilderBase &builder,
+      std::optional<ArrayAttr> privateSyms = targetOp.getPrivateSyms();
+      unsigned numMapVars = targetOp.getMapVars().size();
+      Block &firstTargetBlock = targetRegion.front();
+      auto *blockArgsStart = firstTargetBlock.getArguments().begin();
```
Do only use `auto` when the type is given on the RHS or is too complex.

Dinistro reviewed (commented):
Dropped a few drive by comments. I would delegate to the flang people for a proper review 😅

bhandarkar-pranav commented:
> Dropped a few drive by comments. I would delegate to the flang people for a proper review 😅

@Dinistro  - Thank you for your review

ergawy commented:
Thanks Pranav for the clean and self-contained PR. Can you add a test with a branch or 2 in the `alloc` block?

bhandarkar-pranav commented:
> Thanks Pranav for the clean and self-contained PR. Can you add a test with a branch or 2 in the `alloc` block?

Sounds good, I'll work on such a test.

> I think now you will abandon #105471, right?
Yes, I will.

bhandarkar-pranav committed:
Address review comments

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -3435,6 +3450,59 @@ convertOmpTarget(Operation &opInst, llvm::IRBuilderBase &builder,
+              clonePrivatizer(moduleTranslation, privatizer, &opInst);
+          Region &allocRegion = clonedPrivatizer.getAllocRegion();
+          BlockArgument allocRegionArg = allocRegion.getArgument(0);
+          replaceAllUsesInRegionWith(allocRegionArg, privVar, allocRegion);
```
Instead of replacing all MLIR uses, you could map `allocRegionArg` to the corresponding LLVM value (`moduleTranslation.mapValue`). You can look up this LLVM value from `moduleTranslation`. For example, see how this is done for the `init` region of reductions: https://github.com/llvm/llvm-project/blob/c690d86ca8c0a8f0e7c7fbf2886e7952ea45c728/mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp#L666

Once the MLIR value is mapped to an LLVM value in the `moduleTranslation`, when it converts MLIR operations to LLVM instructions in `inlineConvertOmpRegions` it will use that LLVM value to rewrite all uses of the MLIR value. This avoids needing to modify the MLIR in place.

You can also avoid having to clone the parallel op with a new name by clearing the mapping at the end of each loop iteration (so that it is recreated mapping the block argument to the value for this new iteration).

bhandarkar-pranav replied:
Thanks for the suggestion @tblah. I have made these changes now. Please take a look.

tblah commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -3435,6 +3450,59 @@ convertOmpTarget(Operation &opInst, llvm::IRBuilderBase &builder,
+        }
+      }
+    }
+    builder.restoreIP(codeGenIP);
```
Please could you add a test where there are multiple blocks in the alloc region (if this is allowed by the verifier) to see if this insertion point remains pointing in the right place? We had a lot of bugs of this sort with reductions.

bhandarkar-pranav replied:
Thank you for this suggestion @tblah and @ergawy. Indeed, a multi-block `alloc` region in the privatizer exposed a bug in my code. I have fixed it now. Please take a look. Mind you that the the test I have added has an `alloc` region is artificially multi-block (It needn't be). We'll have real multi-block privatizers when we deal with allocatables which I am working on right now (I have it working using the old approach of inlining in MLIR before converting to LLVMIR. Need to change that code to this newer approach now).

bhandarkar-pranav committed:
Address review comments

 - Add an aritificial multi-blcok alloc region privatizer test in
   mlir/test/Target/LLVMIR/openmp-target-private.mlir
 - Get rid of the need to clone the privatizer and map the
   block argument of the privatizer's alloc region to the llvm value
   corresponding to the host-associated source variable being privatized instead
   of replacing all uses of the block argument the the mlir value
   of the host associated source variable. This obviates the need to
   clone the privatizer.

bhandarkar-pranav committed:
Remove some debug prints

bhandarkar-pranav commented:
@ergawy @tblah  - Updated PR. Please take a look when you get a chance. TIA

ergawy reviewed (approved):
LGTM

tblah reviewed (commented):
Thanks for the updates. The changes look good to me but the CI failure looks relevant

bhandarkar-pranav commented:
> Thanks for the updates. The changes look good to me but the CI failure looks relevant

Yes, the mangled name of the offload function is different in my tree vs the tree built by CI. I'll address this and update the PR. I'll be out of the office on Friday, just FYI

bhandarkar-pranav committed:
Make check for offloaded function name portable across user platforms
